[FEATURE] Add environment configuration section (#357)
TYPO3-Documentation/TYPO3CMS-Tutorial-GettingStarted: Documentation/Installation/EnvironmentConfiguration.rst

variabless, you can access contents of the variables anywhere you need.

.. _environment-helhum-dotenv:

helhum/dotenv-connect
---------------------

You can also use https://github.com/helhum/dotenv-connector/ (via the Packagist
package `helhum/dotenv-connector`) to allow accessing :php:`$_ENV` variables
directly within the Composer autoload process.

This has two nice benefits:

*   You can even set the `TYPO3_CONTEXT` application context environment variable
    through an :file:`.env` file, and no longer need to specify that in your webserver
    configuration (for example, via :file:`.htaccess` or virtual host configuration).
*   You do not need to add and maintain such loading code to your :file:`additional.php`
    file.

The drawback is that you will have an additional dependency on another package, and
glue code that is outside of your own implementation.

.. _environment-phpconfig:

Plain PHP configuration files
=============================

If the concept of requiring a specific file format and their loader dependencies
seems like too much overhead for you, something similar can be achieved

> by including environment-specific PHP files

Reviewer: ```suggestion
by including environment-specific PHP files.
```

For example, you can create a custom file like :file:`system/environment.php` that
will only be placed on your specific target server (and not be kept in your versioning
control system).

..  literalinclude:: Environments/_environment.php
    :language: php
    :caption: config/system/environment.php

This file would also need to be loaded through the additional configuration
workflow (which can be kept in your versioning control system):

..  literalinclude:: Environments/_additional-native.php
    :language: php
    :caption: config/system/additional.php

> You can of course move such a file to a special :file:`Shared/Data/` directory

Reviewer: ```suggestion
Of course, you can move such a file to a special :file:`Shared/Data/` directory
```

.. index:: Environment; Configuration; .env ; dotenv

.. _environment-configuration:

========================
Configuring environments
========================

..  todo: Enhancements

A TYPO3 instance is often used in different contexts that can adapt to your
custom needs:

*   Local development
*   Staging environment
*   Production environment
*   Feature preview environments
*   ...

These can be managed via the same installation by applying different values
for configuration options.

The configuration options can be managed either by an :file:`.env` file or just

> simple :file:`.php` files. Each environment would load the specific

Reviewer: ```suggestion
simple PHP files. Each environment would load the specific
```

> :file:`.env/.php` file, which is usually bound to an

Reviewer: ```suggestion
:file:`.env`/PHP file, which is usually bound to an
```

> :ref:`Application context <t3coreapi:application-context>` (`Development`,

Reviewer: ```suggestion
:ref:`application context <t3coreapi:application-context>` (`Development`,
```

`Production`).

For example, using a :file:`.env` file in your project root, you can define several
variables:

..  literalinclude:: Environments/_example.env
    :language: bash
    :caption: <project-root>/.env

The next step is to retrieve these values in the TYPO3 application bootstrap
process. The best place for this is inside :file:`system/additional.php` (see

> :ref:`t3coreapi:configuration-files`). The PHP code for this could look like:

Reviewer: ```suggestion
:ref:`t3coreapi:configuration-files`). The PHP code for this could looks like:
```

..  literalinclude:: Environments/_additional.php
    :language: php

>     :caption: <config>/system/additional.php

Reviewer: ```suggestion
    :caption: config/system/additional.php
```

Each environment would have its own :file:`.env` file, which is only stored on
the corresponding target system. The development environment file could
be saved as :file:`.env.example` or delivered as the default :file:`.env`
in your project.

..  todo: Should we have a distinct "Versioning" chapter?

It is not recommended to store these :file:`.env` files in your version control
system (e.g. Git). The main reason is that these files usually hold credentials
or other sensitive information.

You should only store environment-specific configuration values in such a
configuration file. Do not use this to manage all the TYPO3 configuration options.
Examples of well-suited configuration options:

* :ref:`t3coreapi:password-policies`
* :ref:`t3coreapi:error-handling-configuration`
* :ref:`t3coreapi:typo3ConfVars_mail`
* :ref:`t3coreapi:typo3ConfVars_sys_encryptionKey`
* :ref:`t3coreapi:security-install-tool`
* Settings, Tokens and URLs to additional services (Redis, Solr, ...)

..  note::

>     Note that the URL of your environment must be configured through

Reviewer: ```suggestion
    The URL of your environment must be configured through
```

    :ref:`site configuration <t3coreapi:sitehandling>` variables, and
    those can actually refer to environment variables as outlined in
    :ref:`t3coreapi:sitehandling-using-env-vars`.

The following sections describe this implementation process in depth.

.. _environment-dotenv:

.env / dotenv files
===================

A central advantage of :file:`.env` files is that environment variables can
also be set in CLI context or injected via Continuous Integration/Deploment

> (CI/CD) systems (GitLab/GitHub) or even Webserver configuration. A central place

Reviewer: ```suggestion
(CI/CD) systems (GitLab/GitHub) or even webserver configuration. A central place
```

> for environment-specific configuration, which many other applications refer to

Reviewer: Is something missing in this sentence? It sounds half-baked.

(and is easy to locate) is also a best practice.

To let your TYPO3 configuration parse keys and values stored of such a file,
you need a library like https://github.com/symfony/dotenv/ or
https://github.com/vlucas/phpdotenv/ and parse it in your :file:`system/additional.php`

You can require these libraries through Packagist/Composer or by including the dependency

> in your project manually (legacy installation).

Reviewer: I think we should not mention legacy here. Why using Composer in a legacy installation for .env handling but not for TYPO3?

Example for `symfony/dotenv`:

..  literalinclude:: Environments/_dotenv-symfony.php
    :language: php
    :caption: <config>/system/additional.php

Example for `vlucas/phpdotenv`:

..  literalinclude:: Environments/_dotenv-vlucas.php
    :language: php
    :caption: <config>/system/additional.php

> Once this code has loaded the :php:`$_ENV` variables from the specified file,

Reviewer: maybe better:

Once this code has loaded the content from .env file into $_ENV variables ...

you can access contents of the variables anywhere you need.

.. _environment-helhum-dotenv:

helhum/dotenv-connect
---------------------

You can also use https://github.com/helhum/dotenv-connector/ (via the Packagist
package `helhum/dotenv-connector`) to allow accessing :php:`$_ENV` variables
directly within the Composer autoload process.

This has two nice benefits:

*   You can even set the `TYPO3_CONTEXT` application context environment variable
    through an :file:`.env` file, and no longer need to specify that in your webserver

>     configuration (e.g. via `.htaccess` or VirtualHost-Configuration).

Reviewer: ```suggestion
    configuration (for example, via :file:`.htaccess` or virtual host configuration).
```

*   You do not need to add and maintain such loading code to your :file:`additional.php`
    file.

> The drawback is that you'll have an additional dependency on another package, and

Reviewer: ```suggestion
The drawback is that you will have an additional dependency on another package, and
```

we want to avoid such shortcuts.

glue code that is outside of your own implementation.

.. _environment-phpconfig:

Plain PHP configuration files
=============================

If the concept of requiring a specific file format and their loader dependencies
seems like too much overhead for you, something similar can be achieved

> by including environment-specific :php:`.php` files

Reviewer: ```suggestion
by including environment-specific PHP files
```

For example, you can create a custom file like :file:`system/environment.php` that
will only be placed on your specific target server (and not be kept in your versioning
control system).

..  literalinclude:: Environments/_environment.php
    :language: php
    :caption: <config>/system/environment.php

This file would also need to be loaded through the additional configuration
workflow:

> ..  literalinclude:: Environments/_additional-native.php

Reviewer: The content is the same is in _environment.php, but should load the _environment.php AFAIU.

Author: Thanks, clarified. I made a boo-boo there when changing files, because additional-native.php is versioned and loads environment.php (I thought that was a bug when I reviewd this)

    :language: php
    :caption: <config>/system/additional.php

You can of course move such a file to a special :file:`Shared/Data/` directory
(see :ref:`deploytypo3`), as long as you take care the file is outside
your public web root directory scope.

The file :file:`additional.php` can still contain custom changes that shall
be applied to every environment of yours, and that is not managed through
:file:`settings.php`.

..  hint::
    The file :file:`settings.php` is used by TYPO3 to store changes made through
    the GUI of the backend. :file:`additional.php` always has the higher
    priority, so configuration values there will overwrite the GUI configuration.

> ..  tip::

Reviewer: Like discussed, please remove this part as it may entangle users and this section should be more about a best practise.

    You can utilize the defined TYPO3
    :ref:`Application context <t3coreapi:application-context>` to dynamically load
    (and enhance) context-specific configuration files, like:

    ..  literalinclude:: Environments/_additional-context.php
        :language: php
        :caption: <config>/system/additional.php

.. _environment-production-settings:

> Production Settings

Reviewer: Hmm, maybe move this to a dedicated chapter?

Author: (Will move this to a dedicated page, yes)

===================

To ensure a secure installation of TYPO3 on a production server, the following settings need to be set:

- :guilabel:`Admin Tools > Settings > Configuration Presets` The "Live" preset has to be chosen to make sure no debug output is displayed.
  When using environment specific configurations, the recommended way is to specifically set the values for
  error/debugging configuration values instead of presets, like:

>     ..  code::

Reviewer: This won't work ;-)

Author: Fixed markup
